pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	if cidr.ClusterID() == 0 {
+
+		ipNet := cidr.AsIPNet()
+		ones, _ := ipNet.Mask.Size()
+
+		if (ipNet.IP.To4() != nil && ones != 32) || (ipNet.IP.To4() == nil && ones != 128) {
+			s.log.Info("CIDR mask not supported", logfields.CIDR, cidr.String())
+		} else {
+			switch modType {
+			case ipcache.Upsert:
+				if oldID != nil {
+					// Remove from the old identity
+					if ips, ok := s.currentIdentityToIp[oldID.ID]; ok {
+						for i, ip := range ips {
+							if ip.Equal(ipNet.IP) {
+								s.currentIdentityToIp[oldID.ID] = append(ips[:i], ips[i+1:]...)
+								break
+							}
+						}
+					}
+				}
+				s.currentIdentityToIp[newID.ID] = append(s.currentIdentityToIp[newID.ID], ipNet.IP)
+
+			case ipcache.Delete:
+				if oldID != nil {
+					if ips, ok := s.currentIdentityToIp[oldID.ID]; ok {
+						for i, ip := range ips {
+							if ip.Equal(ipNet.IP) {
+								s.currentIdentityToIp[oldID.ID] = append(ips[:i], ips[i+1:]...)
+								break
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+// UpdatePolicyRulesLocked updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+func (s *FQDNDataServer) UpdatePolicyRulesLocked(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
+	s.snapshotMutex.Lock()
+	defer s.snapshotMutex.Unlock()
+
+	return nil
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *pb.FQDNMapping) (*pb.UpdateMappingResponse, error) {
+	return &pb.UpdateMappingResponse{
+		Response: pb.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// RunServer starts the Standalone DNS Proxy grpc server on the given port
+func RunServer(port int, server *FQDNDataServer) error {
+	address := fmt.Sprintf("localhost:%d", port)
+	server.log.Info("Starting Standalone DNS Proxy server on: ", logfields.Address, address)
+	lis, err := net.Listen("tcp", address)
```

Comment:
Do we need an actual separate server listening over TCP for this or could we use the UNIX socket we already have (`cilium.sock`)?

Multiplexing gRPC and the REST API over the same socket is not hard: https://github.com/cilium/cilium/pull/28729. 
This PR didn't quite go anywhere as we strictly didn't have a use-case for gRPC at the time. It might be worth having the discussion on whether to have a generic gRPC server to which cells can register handlers easily.

Of course if this needs to be over TCP then it doesn't help ;)

Reply:
There is not any explicit requirement it to be over TCP.  I am happy to be a part of reigniting the https://github.com/cilium/cilium/pull/28729 and using it. 
Also, I don't want to increase the scope of this PR(and the dependent PR that are in queue) as of now if that's okay ?